File: src/203.remove-linked-list-elements/solution.cpp
/*
 * @lc app=leetcode.cn id=203 lang=cpp
 *
 * https://leetcode-cn.com/problems/remove-linked-list-elements/
 */

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// @lc code=start
class Solution {
 public:
  ListNode *removeElements(ListNode *head, int val) {
    ListNode *sentry = new ListNode(0, head);
    ListNode *curr = sentry;

    while (curr) {
      if (curr->next && curr->next->val == val) {
        ListNode *target = curr->next;
        curr->next = target->next;
        delete target;
      } else {
        curr = curr->next;
      }
    }

    ListNode *ans = sentry->next;
    delete sentry;
    return ans;
  }
};
// @lc code=end

File: src/203.remove-linked-list-elements/solution.ts
/*
 * @lc app=leetcode.cn id=203 lang=typescript
 *
 * https://leetcode-cn.com/problems/remove-linked-list-elements/
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start
function removeElements(head: ListNode | null, val: number): ListNode | null {
  const sentry = new ListNode(0, head);
  let current: ListNode | null = sentry;

  while (current) {
    if (current.next && current.next.val === val) {
      current.next = current.next.next;
    } else {
      current = current.next;
    }
  }

  return sentry.next;
}
// @lc code=end

export default removeElements;
